File: Concurrent Programming Lecture Notes/07/07/counter.erl
-module(counter).
-export([start/1, dec/1, inc/1, value/1]).

start(N) ->
  spawn(fun() -> loop(N) end).

inc(Pid) ->
  Pid ! inc.

dec(Pid) ->
  Pid ! dec.

value(Pid) ->
  Pid ! {value, self()},
  receive
    X -> X
  end.

loop(N) ->
  receive
    inc ->
      loop(N + 1);
    dec ->
      loop(N - 1);
    {value, From} ->
      From ! N,
      loop(N)
  end.

File: Concurrent Programming Lecture Notes/06/counter.ex
defmodule Counter do
  def start() do
    :global.register_name(__MODULE__, spawn(fn -> loop(0) end))
  end

  def inc() do
    send(:global.whereis_name(__MODULE__), {:inc, self()})
    receive do
      v -> v
    end
  end

  defp loop(value) do
    receive do
      {:inc, from} ->
        send(from, value + 1)
    end
    loop(value + 1)
  end
end

